templates/ci/flake.nix

```diff
@@ -0,0 +1,37 @@
+{
+  description = "CI Example";
+
+  inputs.nixpkgs.url = "github:nixos/nixpkgs/nixpkgs-unstable";
+  inputs.flake-utils.url = "github:numtide/flake-utils";
+  inputs.nix-utils = {
+    url = "github:iknow/nix-utils";
+    inputs.nixpkgs.follows = "nixpkgs";
+  };
+
+  outputs = { self, nixpkgs, flake-utils, nix-utils }:
+    flake-utils.lib.eachDefaultSystem (system:
+      let pkgs = nixpkgs.legacyPackages.${system}; in
+      {
+        packages.default = pkgs.hello;
+      }
+    ) // (
+      let
+        system = "x86_64-linux";
```

Comment:
I'm still not quite sure how we're going to make this cross arch so just setting it to x64 linux for now.

Reply:
Opinions welcome on https://github.com/iknow/nix-utils/pull/3. Maybe we need a builder that drives an image manifest factory by system?

```
makeSimpleMultiArchImage "my-image" systems (system: {
  config.Entrypoint = [ "${self.packages.${system}.foo/bin/foo" ];
})
```

Reply:
That looks reasonable, but what arch is `makeSimpleMultiArchImage` running in? We can't just use `currentSystem` for that right?

Also, what about tests and `useHostStore`?

Reply:
> That looks reasonable, but what arch is `makeSimpleMultiArchImage` running in? We can't just use `currentSystem` for that right?

Isn't this question already answered by nix-utils/flake.nix? They're exported under `utils.${system}.oci`. Though I admit there's no much scope for using the "local" system, for example, if we had aarch64-linux jenkins nodes.

> Also, what about tests and `useHostStore`?

Tests need to run on the host platform, this seems pretty unavoidable. We could probably export a jenkins label selector as part of the test metadata phase, and then `nix copy` the tests to the selected node (somehow?). `useHostStore` _shouldn't_ be interesting, since the all the dependencies should be in the test configuration's closure.

Reply:
> I'm still not quite sure how we're going to make this cross arch so just setting it to x64 linux for now.

Just to be clear, I think this is fine for now, but I do like thinking about the multiple architecture images.

Reply:
> Isn't this question already answered by nix-utils/flake.nix? They're exported under `utils.${system}.oci`. Though I admit there's no much scope for using the "local" system, for example, if we had aarch64-linux jenkins nodes.

I was more asking about who builds the image index manifest itself since someone has to build it. Do we just take one of the `systems`?

So `utils.${builtins.head systems}.makeSimpleMultiArchImage ...`?

Reply:
I was doing absolutely nothing clever
```nix
{
   # The system doing the building, with the expectation that builds
   # are local.
   buildSystem ? "x86_64-linux",

   # System to build for. Note this is *not* cross compilation.
   exportedSystems ? [ "x86_64-linux" "aarch64-linux" ],
 }:
let
   buildPkgs = import ./nix/pinned-nixpkgs.nix { system = buildSystem; };
   eikaiwaBuildPkgs = import ./default.nix { pkgs = buildPkgs; };
   inherit (eikaiwaBuildPkgs) ociTools;
in
# ...
```